delta/plugins/storage/src/test/scala/ch/epfl/bluebrain/nexus/delta/plugins/storage/files/batch/BatchCopySuite.scala

```diff
@@ -0,0 +1,293 @@
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.Storage.{DiskStorage, RemoteDiskStorage, S3Storage}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.model.{Storage, StorageStatEntry, StorageType}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.StorageFileRejection.CopyFileRejection.{DifferentStorageTypes, SourceFileTooLarge, TotalCopySizeTooLarge, UnsupportedOperation}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.disk.{DiskCopyDetails, DiskStorageCopyFiles}
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.remote.RemoteDiskStorageCopyFiles
+import ch.epfl.bluebrain.nexus.delta.plugins.storage.storages.operations.remote.client.model.RemoteDiskCopyDetails
+import ch.epfl.bluebrain.nexus.delta.rdf.Vocabulary.nxv
+import ch.epfl.bluebrain.nexus.delta.sdk.acls.model.AclAddress
+import ch.epfl.bluebrain.nexus.delta.sdk.acls.{AclCheck, AclSimpleCheck}
+import ch.epfl.bluebrain.nexus.delta.sdk.error.ServiceError.AuthorizationFailed
+import ch.epfl.bluebrain.nexus.delta.sdk.identities.model.Caller
+import ch.epfl.bluebrain.nexus.delta.sdk.model.{IdSegment, IdSegmentRef, Tags}
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.Identity.User
+import ch.epfl.bluebrain.nexus.delta.sourcing.model.{ProjectRef, ResourceRef}
+import ch.epfl.bluebrain.nexus.testkit.Generators
+import ch.epfl.bluebrain.nexus.testkit.mu.NexusSuite
+
+import java.util.UUID
+import scala.collection.mutable.ListBuffer
+
+class BatchCopySuite extends NexusSuite with StorageFixtures with Generators with FileFixtures with FileGen {
+
+  private val sourceFileDescUuid         = UUID.randomUUID()
+  implicit private val fixedUUidF: UUIDF = UUIDF.fixed(sourceFileDescUuid)
+
+  private val sourceProj       = genProject()
+  private val sourceFileId     = genFileId(sourceProj.ref)
+  private val source           = CopyFileSource(sourceProj.ref, NonEmptyList.of(sourceFileId))
+  private val storageStatEntry = StorageStatEntry(files = 10L, spaceUsed = 5L)
+  private val stubbedFileAttr  = NonEmptyList.of(attributes(genString()))
+
+  test("successfully perform disk copy") {
+    val events                         = ListBuffer.empty[Event]
+    val (sourceFileRes, sourceStorage) = genFileResourceAndStorage(sourceFileId, sourceProj.context, diskVal)
+    val (user, aclCheck)               = userAuthorizedOnProjectStorage(sourceStorage.value)
+
+    val batchCopy                = mkBatchCopy(
+      fetchFile = stubbedFetchFile(sourceFileRes, events),
+      fetchStorage = stubbedFetchStorage(sourceStorage, events),
+      aclCheck = aclCheck,
+      stats = stubbedStorageStats(storageStatEntry, events),
+      diskCopy = stubbedDiskCopy(stubbedFileAttr, events)
+    )
+    val destStorage: DiskStorage = genDiskStorage()
+
+    batchCopy.copyFiles(source, destStorage)(caller(user)).map { obtained =>
+      val obtainedEvents = events.toList
```

Comment:
I wonder if an abstraction around the events list might be a bit clearer

Reply:
There is actually a really nice pure FP abstraction for this kinda of testing (no mutable state), but it relies on tagless final / monad transformers so not a good fit here. I'll think about how to make it clearer tho 👍

Reply:
I don't think the mutable state is the problem, it just needs to read better. I was thinking just a tiny class / enclosure